File: engine/src/main/java/com/terralite/engine/terrain/SparseBlockStorage.java
package com.terralite.engine.terrain;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public final class SparseBlockStorage implements BlockStorage {
    private final Map<BlockPos, BlockState> blocks = new LinkedHashMap<>();

    @Override
    public BlockState get(BlockPos pos) {
        return blocks.getOrDefault(Objects.requireNonNull(pos, "pos"), BlockState.AIR);
    }

    @Override
    public void set(BlockPos pos, BlockState state) {
        Objects.requireNonNull(pos, "pos");
        Objects.requireNonNull(state, "state");
        if (state.isAir()) {
            blocks.remove(pos);
            return;
        }
        blocks.put(pos, state);
    }

    @Override
    public BlockState remove(BlockPos pos) {
        BlockState removed = blocks.remove(Objects.requireNonNull(pos, "pos"));
        return removed == null ? BlockState.AIR : removed;
    }

    @Override
    public boolean contains(BlockPos pos) {
        return blocks.containsKey(Objects.requireNonNull(pos, "pos"));
    }

    @Override
    public Collection<BlockPos> positions() {
        return List.copyOf(blocks.keySet());
    }

    @Override
    public int size() {
        return blocks.size();
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/terrain/SparseBlockStorage.kt
package com.terralite.engine.terrain

class SparseBlockStorage : BlockStorage {
    private val blocks: MutableMap<BlockPos, BlockState> = LinkedHashMap()

    override fun get(pos: BlockPos): BlockState = blocks.getOrDefault(pos, BlockState.AIR)

    override fun set(pos: BlockPos, state: BlockState) {
        if (state.isAir()) {
            blocks.remove(pos)
        } else {
            blocks[pos] = state
        }
    }

    override fun remove(pos: BlockPos): BlockState = blocks.remove(pos) ?: BlockState.AIR

    override fun contains(pos: BlockPos): Boolean = blocks.containsKey(pos)

    override fun positions(): Collection<BlockPos> = blocks.keys.toList()

    override fun size(): Int = blocks.size
}
